### DISCUSSION TITLE
Does a leading "_" in a module name mean "internal"?

### DISCUSSION THREAD

**Discussion by wayneadams** [aeon-toolkit/aeon#1429] [Q&A] [answered]
Is there a reason why many of the modules that have leading "internal" underscores? When I see "_", I assume that that I should not use the things within. 

For example:
- [from_3d_numpy_to_2d_array](https://github.com/aeon-toolkit/aeon/blob/136d218817f4bea5da8f6283528083f7a37d47ce/aeon/datatypes/_panel/_convert.py#L173) seems to be a public function, but its parent is `_convert.py`
- [WeightedEnsembleClassifier](https://github.com/aeon-toolkit/aeon/blob/136d218817f4bea5da8f6283528083f7a37d47ce/aeon/classification/compose/_ensemble.py#L15) is certainly a public class, but its parent is `_ensemble.py`

But, maybe there's some other meaning that I don't know about. Do the leading underscores for modules mean "internal"?

**Comment by baraline** [comment, marked as answer]
I think the idea behind this design is that you  should import/use things exposed from __init__ files, which themselves will import functions from the modules with underscores. 
This allons to define whats part of the "public" API, which can be subject to deprecation etc..
Importing straight from _***.py module is not very practical/pythonic either.

**Reply by wayneadams** [reply]
That makes 100% sense! I forgot about `__init__` files. I like the idea of collecting all the important things in just a few places.

Thanks for the quick reply!
